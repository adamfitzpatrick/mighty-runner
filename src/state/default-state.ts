import { Character, PersonalData, CharacterIdentifier, Attributes, Effect, GearItem, Pic } from '@models'

export interface AppState {
  characters: Character[] | null
  activeCharacter: CharacterIdentifier | null
  pic: Pic | null
  personalData: PersonalData | null
  attributes: Attributes | null
  gear: GearItem[] | null
  effects: Effect[] | null
  apiError: boolean
}

export const defaultState: AppState = {
  characters: null,
  pic: null,
  activeCharacter: null,
  personalData: null,
  attributes: null,
  gear: null,
  effects: null,
  apiError: false
}

export default defaultState
