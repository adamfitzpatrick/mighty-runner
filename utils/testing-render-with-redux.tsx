import * as React from 'react'
import {
  createStore,
  combineReducers,
  AnyAction,
  Store,
  Dispatch,
  applyMiddleware
} from 'redux'
import { Provider } from 'react-redux'
import { render, RenderResult } from '@testing-library/react'
import * as Models from '@models'
import { AppState } from '../src/state/default-state'

export type RenderWithRedux =
  RenderResult &
  {
    store: Store<AppState, AnyAction>,
    dispatchSpy: jest.Mock<Dispatch<AnyAction>>
  }

const reducer = combineReducers({
  characters: (state: Models.Character[] | null = null) => state,
  activeCharacter: (state: Models.CharacterIdentifier | null = null) => state,
  personalData: (state: Models.PersonalData | null = null) => state,
  attributes: (state: Models.Attributes | null = null) => state,
  gear: (state: Models.GearItem[] | null = null) => state,
  effects: (state: Models.Effect[] | null = null) => state
})

const getSpyMiddleware = (dispatchSpy: jest.Mock<Dispatch<AnyAction>>) => () => {
  return (next: Dispatch<AnyAction>) => (action: AnyAction) => {
    dispatchSpy(action)
    next(action)
  }
}

export function renderWithRedux (
  ui: JSX.Element,
  initialState: AppState
): RenderWithRedux {
  const dispatchSpy: jest.Mock<Dispatch<AnyAction>> = jest.fn()
  const spyMiddleware = getSpyMiddleware(dispatchSpy)
  const store = createStore(reducer, initialState, applyMiddleware(spyMiddleware))

  return {
    ...render(<Provider store={store}>{ ui }</Provider>),
    store,
    dispatchSpy
  }
}
