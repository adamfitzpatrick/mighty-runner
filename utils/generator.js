const fs = require('fs')
const path = require('path')
const yargs = require('yargs')

if (!yargs.argv.container && !yargs.argv.component) {
  throw new Error('no component or container name provided')
}

const container = yargs.argv.container && yargs.argv.container.replace(/^[a-z]/, a => a.toUpperCase())
const component = yargs.argv.component && yargs.argv.component.replace(/^[a-z]/, a => a.toUpperCase())

let thing = yargs.argv.component
if (container) {
  thing = container
}
const lower = thing.replace(/^[A-Z]/, a => a.toLowerCase()).replace(/[A-Z]/g, a => `-${a.toLowerCase()}`)
thing = thing.replace(/^[a-z]/, a => a.toUpperCase()).replace(/-([A-Za-z0-9])/g, a => a.slice(1).toUpperCase())
scssClass = thing.replace(/^[A-Z]/, a => a.toLowerCase())

const thingTs = `import * as React from 'react'

import * as styles from './${lower}.scss'

interface Props {}

export default function ${thing} (props: Props) {
  return <div>${thing}</div>
}
`

const thingTestTs = `import * as React from 'react'
import { cleanup, render } from '@testing-library/react'
import { renderWithRedux, RenderWithRedux } from '../../../utils/testing-render-with-redux'
import ${thing} from '.'

describe('${thing} ${ container ? 'container' : 'component'}', () => {
  test('should do nothing', () => {
    expect(true).toEqual(true)
  })
})
`

const indexTs = `export { default } from './${lower}'
`

const scss = `.${scssClass} {
  display: block;
}
`

const scssD = `export const ${scssClass}: string
`

let dest = path.resolve(process.cwd(), 'src', 'components', lower)
if (container) {
  dest = path.resolve(process.cwd(), 'src', 'containers', lower)
}

fs.mkdirSync(dest)
fs.writeFileSync(path.resolve(dest, `${lower}.tsx`), thingTs)
fs.writeFileSync(path.resolve(dest, `${lower}.test.tsx`), thingTestTs)
fs.writeFileSync(path.resolve(dest, 'index.ts'), indexTs)
fs.writeFileSync(path.resolve(dest, `${lower}.scss`), scss)
fs.writeFileSync(path.resolve(dest, `${lower}.scss.d.ts`), scssD)
