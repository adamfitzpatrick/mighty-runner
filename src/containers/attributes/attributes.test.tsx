import * as React from 'react'
import { renderWithRedux, RenderWithRedux } from '../../../utils/testing-render-with-redux'
import { fireEvent, cleanup } from '@testing-library/react'
import Attributes from '.'
import * as Models from '@models'
import { AppState } from '@state/default-state'

describe('Attributes container', () => {
  let attributes: Models.Attributes
  let effects: Models.Effect[]
  let appState: AppState
  let sut: RenderWithRedux

  beforeEach(() => {
    attributes = {
      body: {
        id: 'body',
        name: 'Body',
        shortName: 'bod',
        description: 'd',
        asEffectTarget: [ 'attributes', 'body' ],
        value: { initial: 0, chargen: 6 },
        maximum: 6
      },
      agility: {
        id: 'agility',
        name: 'Agility',
        shortName: 'agi',
        description: 'd',
        asEffectTarget: [ 'attributes', 'agility' ],
        value: { initial: 0, chargen: 6 },
        maximum: 6
      },
      reaction: {
        id: 'reaction',
        name: 'Reaction',
        shortName: 'rea',
        description: 'd',
        asEffectTarget: [ 'attributes', 'reaction' ],
        value: { initial: 0, chargen: 6 },
        maximum: 6
      },
      strength: {
        id: 'strength',
        name: 'Strength',
        shortName: 'str',
        description: 'd',
        asEffectTarget: [ 'attributes', 'strength' ],
        value: { initial: 0, chargen: 6 },
        maximum: 6
      },
      logic: {
        id: 'logic',
        name: 'Logic',
        shortName: 'log',
        description: 'd',
        asEffectTarget: [ 'attributes', 'logic' ],
        value: { initial: 0, chargen: 6 },
        maximum: 6
      },
      willpower: {
        id: 'willpower',
        name: 'Willpower',
        shortName: 'wil',
        description: 'd',
        asEffectTarget: [ 'attributes', 'willpower' ],
        value: { initial: 0, chargen: 6 },
        maximum: 6
      },
      intuition: {
        id: 'intuition',
        name: 'Intuition',
        shortName: 'int',
        description: 'd',
        asEffectTarget: [ 'attributes', 'intuition' ],
        value: { initial: 0, chargen: 6 },
        maximum: 6
      },
      charisma: {
        id: 'charisma',
        name: 'Charisma',
        shortName: 'cha',
        description: 'd',
        asEffectTarget: [ 'attributes', 'charisma' ],
        value: { initial: 0, chargen: 6 },
        maximum: 6
      },
      essence: {
        id: 'essence',
        name: 'Essence',
        shortName: 'ess',
        description: 'd',
        asEffectTarget: [ 'attributes', 'essence' ],
        value: { initial: 0, chargen: 6 },
        maximum: 6
      }
    }
    effects = [{
      id: '1',
      name: 'n',
      description: 'd',
      target: [ 'attributes', 'agility' ],
      value: 1,
      active: true
    }]
    appState = {
      attributes,
      effects
    } as AppState
    sut = renderWithRedux(<Attributes />, appState)
  })

  test('should render correctly', () => {
    expect(sut.container.innerHTML).toMatchSnapshot()
  })

  test('should not throw an error if the redux state is missing', () => {
    cleanup()
    delete appState.attributes
    sut = renderWithRedux(<Attributes />, appState)
    expect(sut.container.innerHTML).toMatchSnapshot()
  })

  describe('editing', () => {
    let editButton

    beforeEach(() => {
      editButton = sut.getAllByTestId('Edit.button.component')[0]
      fireEvent.click(editButton)
    })

    test('should show the edit component when the edit button is clicked', () => {
      expect(sut.getByTestId('Body.edit-item.component')).toBeDefined()
    })

    test('should update local state on edit-item field changes', () => {
      const input = sut.getByTestId('Maximum.input.component') as HTMLInputElement
      fireEvent.change(input, { target: { value: '3' } })
      expect(input.value).toEqual('3')
    })

    test('should update the attributes when editing is finished', () => {
      const doneButton = sut.getByTestId('Done.button.component')
      fireEvent.click(doneButton)
      expect(sut.dispatchSpy.mock.calls[0][0].type).toEqual('UPDATE_ATTRIBUTE')
    })
  })

})
