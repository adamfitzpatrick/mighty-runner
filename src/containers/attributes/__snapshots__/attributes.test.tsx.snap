// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Attributes container should not throw an error if the redux state is missing 1`] = `""`;

exports[`Attributes container should render correctly 1`] = `"<div><h2>Attributes</h2><div><div><h3>Body (bod)</h3><div><span>Effective Value:<span data-testid=\\"attributes.body.effective-stat.container\\">6</span></span><button data-testid=\\"Edit.button.component\\">Edit</button></div></div></div><div><div><h3>Agility (agi)</h3><div><span>Effective Value:<span data-testid=\\"attributes.agility.effective-stat.container\\">7</span></span><button data-testid=\\"Edit.button.component\\">Edit</button></div><div data-testid=\\"non-editable-effect.component\\"><span>Name: n</span><label><span>Active</span><input data-testid=\\"Active.input.component\\" type=\\"checkbox\\" value=\\"\\" checked=\\"\\"></label></div></div></div><div><div><h3>Reaction (rea)</h3><div><span>Effective Value:<span data-testid=\\"attributes.reaction.effective-stat.container\\">6</span></span><button data-testid=\\"Edit.button.component\\">Edit</button></div></div></div><div><div><h3>Strength (str)</h3><div><span>Effective Value:<span data-testid=\\"attributes.strength.effective-stat.container\\">6</span></span><button data-testid=\\"Edit.button.component\\">Edit</button></div></div></div><div><div><h3>Willpower (wil)</h3><div><span>Effective Value:<span data-testid=\\"attributes.willpower.effective-stat.container\\">6</span></span><button data-testid=\\"Edit.button.component\\">Edit</button></div></div></div><div><div><h3>Logic (log)</h3><div><span>Effective Value:<span data-testid=\\"attributes.logic.effective-stat.container\\">6</span></span><button data-testid=\\"Edit.button.component\\">Edit</button></div></div></div><div><div><h3>Intuition (int)</h3><div><span>Effective Value:<span data-testid=\\"attributes.intuition.effective-stat.container\\">6</span></span><button data-testid=\\"Edit.button.component\\">Edit</button></div></div></div><div><div><h3>Charisma (cha)</h3><div><span>Effective Value:<span data-testid=\\"attributes.charisma.effective-stat.container\\">6</span></span><button data-testid=\\"Edit.button.component\\">Edit</button></div></div></div></div>"`;
